src/audio/drc/drc.c

```diff
@@ -0,0 +1,483 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Google LLC. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <sof/audio/component.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/drc/drc.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/math/decibels.h>
+#include <sof/math/numbers.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/eq.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_drc;
+
+/* TODO: get the real UUID */
+/* 5150c0e7-27f9-4ec8-8351-c705b642d12f */
+DECLARE_SOF_RT_UUID("drc", drc_uuid, 0x5150c0e7, 0x27f9, 0x4ec8,
+		    0x83, 0x51, 0xc7, 0x05, 0xb6, 0x42, 0xd1, 0x2f);
+
+DECLARE_TR_CTX(drc_tr, SOF_UUID(drc_uuid), LOG_LEVEL_INFO);
+
+static inline void drc_reset_state(struct drc_state *state)
+{
+	int i;
+
+	rfree(state->pre_delay_buffers[0]);
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; ++i) {
+		state->pre_delay_buffers[i] = NULL;
+	}
+
+	state->detector_average = 0;
+	state->compressor_gain = Q_CONVERT_FLOAT(1.0f, 30);
+
+	state->last_pre_delay_frames = DRC_DEFAULT_PRE_DELAY_FRAMES;
+	state->pre_delay_read_index = 0;
+	state->pre_delay_write_index = DRC_DEFAULT_PRE_DELAY_FRAMES;
+
+	state->envelope_rate = 0;
+	state->scaled_desired_gain = 0;
+
+	state->processed = 0;
+
+	state->max_attack_compression_diff_db = INT32_MIN;
+}
+
+static inline int drc_init_pre_delay_buffers(struct drc_state *state,
+					     size_t sample_bytes,
+					     int channels)
+{
+	size_t bytes_per_channel = sample_bytes * DRC_MAX_PRE_DELAY_FRAMES;
+	size_t bytes_total = bytes_per_channel * channels;
+	int i;
+
+	/* Allocate pre-delay (lookahead) buffers */
+	state->pre_delay_buffers[0] =
+		rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, bytes_total);
+	if (!state->pre_delay_buffers[0])
+		return -ENOMEM;
+
+	memset(state->pre_delay_buffers[0], 0, bytes_total);
+
+	for (i = 1; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
+			state->pre_delay_buffers[i - 1] + bytes_per_channel;
```

Comment:
I realized that this workes because pre_delay_buffers[] is int8_t *. The data is really int16_t it seems so it's type casted when used. If the data is always int16_t maybe the definition of pre_delay_buffers[] and this pointer arithmetic should change? But if it's stream dependent varying type this is probably best.

Reply:
Yeah, it is stream dependent varying type. Although currently I only implemented S16 case.